compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/reshaping/SamplingSkewReshapingPass.java

```diff
@@ -0,0 +1,126 @@
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Optimizes the PartitionSet property of shuffle edges to handle data skews using the SamplingVertex.
+ *
+ * This pass effectively partitions the IRDAG by non-oneToOne edges, clones each subDAG partition using SamplingVertex
+ * to process sampled data, and executes each cloned partition prior to executing the corresponding original partition.
+ *
+ * Suppose the IRDAG is partitioned into two sub-DAGs as follows:
+ * P1 - P2
+ *
+ * Then, this pass will produce something like:
+ * P1' - P1 - P2
+ *          - P2' - P2
+ * where Px' consists of SamplingVertex objects that clone the execution of Px.
+ *
+ * For each Px' this pass also inserts a MessageBarrierVertex, to use its data statistics for dynamically optimizing
+ * the execution behaviors of Px.
+ */
+@Requires(CommunicationPatternProperty.class)
+public final class SamplingSkewReshapingPass extends ReshapingPass {
+  private static final Logger LOG = LoggerFactory.getLogger(SamplingSkewReshapingPass.class.getName());
+  private static final float SAMPLE_RATE = 0.1f;
+
+  /**
+   * Default constructor.
+   */
+  public SamplingSkewReshapingPass() {
+    super(SamplingSkewReshapingPass.class);
+  }
+
+  @Override
+  public IRDAG apply(final IRDAG dag) {
+    dag.topologicalDo(v -> {
+      for (final IREdge e : dag.getIncomingEdgesOf(v)) {
+        if (CommunicationPatternProperty.Value.Shuffle.equals(
+          e.getPropertyValue(CommunicationPatternProperty.class).get())) {
+          // Compute the partition and its source vertices
+          final IRVertex shuffleWriter = e.getSrc();
+          final Set<IRVertex> partitionAll = recursivelyBuildPartition(shuffleWriter, dag);
+          final Set<IRVertex> partitionSources = partitionAll.stream().filter(vertexInPartition ->
+            !dag.getIncomingEdgesOf(vertexInPartition).stream()
+              .map(Edge::getSrc)
+              .anyMatch(partitionAll::contains)
+          ).collect(Collectors.toSet());
+
+          // Insert sampling vertices.
+          final Set<SamplingVertex> samplingVertices = partitionAll
+            .stream()
+            .map(vertexInPartition -> new SamplingVertex(vertexInPartition, SAMPLE_RATE))
+            .collect(Collectors.toSet());
+          dag.insert(samplingVertices, partitionSources);
+
+          // Insert the message vertex.
+          // We first obtain a clonedShuffleEdge to analyze the data statistics of the shuffle outputs of
+          // the sampling vertex right before shuffle.
+          final SamplingVertex rightBeforeShuffle = samplingVertices.stream()
+            .filter(sv -> sv.getOriginalVertex().equals(e.getSrc()))
+            .findFirst()
+            .orElseThrow(() -> new IllegalStateException());
+          final IREdge clonedShuffleEdge = rightBeforeShuffle.getCloneOfOriginalEdge(e);
+
+          final KeyExtractor keyExtractor = e.getPropertyValue(KeyExtractorProperty.class).get();
```

Comment:
We need to add control edges from the message aggregation vertex to the `partitionSources` instead of the vertex that receives the original shuffle edge.

For the example DAG that is partitioned into two sub-DAGs as follows: P1 -(shuffle)- P2,
the expected outcome looks like P1' -(o2o)- MCV -(shuffle)- MAV -(control)- P1 -(shuffle)- p2.

This is because that we **must** optimze the partitioning way of the target shuffle edge before the execution of P1.

Also, the P1' and message collection vertex must be in a single stage. If not, the whole intermediate data will be duplicated.

Reply:
Thanks for bringing this up!

(1) Regarding pipelining MessageBarrierVertex within a single stage with parent sampling vertices:

I've changed the semantics of insert() to use SamplingVertex(NewVertex) instead of the NewVertex, if an existing vertex that the NewVertex will connect to is a SamplingVertex. I think this is a reasonable assumption as new vertices that consume outputs from sampling vertices will process a subset of data anyways, and no such new vertex will  reach the original DAG except via control edges. With this change Nemo is able to pipeline the MessageBarrierVertex (wrapped inside a SamplingVertex), avoiding duplicate data materialization.

(2) Regarding connecting the message aggregation vertex to the partition Sources:

I'd prefer not to do this, at least considering current use cases we have. 

Here's the physical DAG diagram of PerKeyMedianITCase#testLargeShuffleSamplingSkew  (including the fix for (1))
https://nemo.snuspl.snu.ac.kr:50443/nemo-dag-out/7a1c136ac24f427ebb4c34a43712da3f.svg

In the diagram the ScheduleGroup property is set such that the sampling partition does always execute prior to the original partition. In particular the ordering ScheduleGroup0(Stage3+Stage5) ==> ScheduleGroup1(Stage4+Stage6) is enforced although Stage4 ==> Stage6 is PUSH (which makes sense when considering each schedule group as a big vertex). The sampling should also happen prior to the execution of the original partition when Stage4 ==> Stage6 is PULL as well, although the schedule groups may differ in this case. I think this shows that a sequence of insert(samplingVertex) and insert(messageVertex) captures our intention fairly well.

I did write some code to try to 'extend' the control edges from (sampling vertices) to (existing vertices), by adding new control edges from (new vertices that connect to sampling vertices) to (existing vertices) upon each insert(). However, I ultimately I felt that this approach complicates the code quite a bit, and reverted the code back to the current approach which I think works for the current use cases.

Reply:
For (2), I understood why the current version of the pass works normally.
However, scheduling indirectly depending on the logic of scheduling group is at a risk. (The order of scheduling according to the scheduling group is implicit and might be changed.)
If it is not simple to add the control dependency as I mentioned, please create an issue about it and mark as TODO.

Reply:
Agreed. I've filed the JIRA: https://issues.apache.org/jira/browse/NEMO-343